### PR TITLE
feat(connect-web): automatic fallback to core in popup

### LINKED ISSUE #12488: Add webUSB as a secondary transport option to connect-web package
In `trezor/connect-web`:

if **Trezor Bridge** is running, use **core in iframe**

if **Trezor Bridge** is not running and/or **core in iframe** cannot be used, use **core in popup**

### PR DESCRIPTION
**Pull request by martykan**
## Description

Implement automatic fallback for core in popup. 

- New param `coreMode`
    - `auto` - automatically choose the best option based on the environment
    - `iframe` (default) - always use iframe, same way as before
    - `popup` - always use popup,
- Switch to popup on error (`Init_IframeBlocked`, `Init_IframeTimeout`, `Transport_Missing`, …)
- Checking for bridge state is done using a new event `TRANSPORT.GET_INFO`, which will trigger `Transport_Missing` during init if in auto mode

## Related Issue
Resolve #12488

## Followup
- Deprecate `useCoreInPopup` param
- Update documentation
- Use auto as default?

### CODE REVIEW

File: packages/connect-web/src/utils/proxy-event-emitter.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import EventEmitter from 'events';
+
+export default class ProxyEventEmitter implements EventEmitter {
```
Comment (mroz22) [inline packages/connect-web/src/utils/proxy-event-emitter.ts:7]: maybe please add a comment why is this needed, I remember we were discussing this and it made sense back then but after 3 weeks I am completely lost :D

Comment (martykan) [reply packages/connect-web/src/utils/proxy-event-emitter.ts:7]: It can basically combine N event emitters into 1, for the different core in X implementations

File: packages/connect-explorer/src/pages/settings.mdx [hunk 2]
```diff
@@ -52,6 +53,17 @@ export const Settings = () => {
             key: 'useCoreInPopup',
```
Comment (mroz22) [inline packages/connect-explorer/src/pages/settings.mdx:54]: maybe we could get rid of it already? I understand that this requires some work in tests...

File: packages/connect-web/src/utils/proxy-event-emitter.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        return this.eventEmitters[0].listenerCount(eventName, listener);
+    }
+
+    rawListeners(eventName: string | symbol) {
+        return this.eventEmitters[0].rawListeners(eventName);
+    }
+
+    listeners(eventName: string | symbol) {
+        return this.eventEmitters[0].listeners(eventName);
+    }
```
Comment (mroz22) [inline packages/connect-web/src/utils/proxy-event-emitter.ts:94]: these are there only for types compatibility?

**Comment by mroz22** [comment]
this can be probably merged separately? 6252066d173d09fd525fd71ae4601c02ccd87172 also do we have a way how to automatically detect bundling stuff we don't want to bundle? (I know I asked this some time ago but I forgot) :/

File: packages/connect-web/src/utils/proxy-event-emitter.ts (hunk 3 above)
Comment (martykan) [reply packages/connect-web/src/utils/proxy-event-emitter.ts:94]: Mostly yes but some of these methods are actually used.

File: packages/connect-web/src/impl/core-in-iframe.ts [hunk 4]
```diff
@@ -174,6 +173,18 @@ export class CoreInIframe implements ConnectFactoryDependencies {
...
+            console.log('Checking bridge availability'); // DEBUG
+            const { promiseId, promise } = this._messagePromises.create();
+            iframe.postMessage({ id: promiseId, type: TRANSPORT.GET_INFO });
+            const response = await promise;
+            console.log('Bridge availability response', response);
+            if (response.payload === undefined) {
+                throw ERRORS.TypedError('Transport_Missing');
+            }
+        }
+
```
Comment (mroz22) [inline packages/connect-web/src/impl/core-in-iframe.ts]: I am still thinking about the race-condition scenario. not sure if we were able to rule it out or not. the worst thing that can happen is that under mode `auto` we will switch to popup mode even when it is not necessary, right?

File: packages/connect-explorer/src/pages/settings.mdx (hunk 2 above)
Comment (martykan) [reply packages/connect-explorer/src/pages/settings.mdx:54]: Yes, but I think we will actually have to leave it internally, for compatibility with older `connect-webextension` for example

File: packages/connect-web/src/impl/core-in-iframe.ts [hunk 5]
```diff
@@ -193,19 +204,13 @@ export class CoreInIframe implements ConnectFactoryDependencies {
...
                 this._popupManager = this.initPopupManager();
             }
-            this._popupManager.request();
 
             // auto init with default settings
             try {
                 await this.init(this._settings);
             } catch (error) {
                 if (this._popupManager) {
-                    // Catch fatal iframe errors (not loading)
```
Comment (mroz22) [inline packages/connect-web/src/impl/core-in-iframe.ts:203]: cool, so this means that these errors are no longer fatal and we can retry usign a different way

File: packages/connect-web/src/impl/core-in-iframe.ts (hunk 4 above)
Comment (martykan) [reply packages/connect-web/src/impl/core-in-iframe.ts]: Yes, nothing too bad should happen

Comment (martykan) [reply packages/connect-web/src/impl/core-in-iframe.ts]: But I actually changed `transportReconnect` to wait in case of `auto` mode

File: packages/connect-web/src/impl/core-in-popup.ts [hunk 6]
```diff
@@ -130,26 +130,26 @@ export class CoreInPopup implements ConnectFactoryDependencies {
...
+                    },
+                });
+            }
 
-        await this._popupManager.handshakePromise?.promise;
+            await this._popupManager.handshakePromise?.promise;
 
-        // post message to core in popup
-        try {
+            // post message to core in popup
```
Comment (mroz22) [inline packages/connect-web/src/impl/core-in-popup.ts:152]: what is the reason for this change?

File: packages/connect-web/src/utils/proxy-event-emitter.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import EventEmitter from 'events';
+
+/**
+ * ProxyEventEmitter is an EventEmitter that allows to use multiple EventEmitters as one
+ * This is used in connect-web to allow switching between iframe and core-in-popup implementations
+ */
+export default class ProxyEventEmitter implements EventEmitter {
```
Comment (mroz22) [inline packages/connect-web/src/utils/proxy-event-emitter.ts:7]: cc @marekrjpolak 

this is a "utilitish code". at the moment I am not seeing any other usecase for it that would advocate moving this into packages/utils. still wanted to ping you about it if you have any opinion/idea

File: packages/connect/src/data/connectSettings.ts [hunk 8]
```diff
@@ -137,6 +137,13 @@ export const parseConnectSettings = (input: Partial<ConnectSettings> = {}) => {
         settings.useCoreInPopup = input.useCoreInPopup;
     }
 
+    if (
```
Comment (mroz22) [inline packages/connect/src/data/connectSettings.ts:140]: more like a note for me: 
btw defaults are set only on the 'front-end' level?

File: packages/connect-web/src/utils/proxy-event-emitter.ts (hunk 7 above)
Comment (marekrjpolak) [reply packages/connect-web/src/utils/proxy-event-emitter.ts:7]: I'd leave it as it is for now, and potentially move it to `@trezor/utils` as soon as we want to use something similar for the 2nd time.

File: packages/connect-web/src/impl/core-in-popup.ts (hunk 6 above)
Comment (martykan) [reply packages/connect-web/src/impl/core-in-popup.ts:152]: We want to catch all errors and return `createErrorMessage`

File: packages/connect-web/src/utils/proxy-event-emitter.ts (hunk 3 above)
Comment (mroz22) [reply packages/connect-web/src/utils/proxy-event-emitter.ts:94]: ok pleease add a comment that they have everything the same

File: packages/connect-web/src/impl/core-in-iframe.ts [hunk 9]
```diff
@@ -174,6 +173,18 @@ export class CoreInIframe implements ConnectFactoryDependencies {
 
         await iframe.init(this._settings);
 
+        if (this._settings.coreMode === 'auto') {
+            // Checking bridge availability
+            console.log('Checking bridge availability'); // DEBUG
```
Comment (mroz22) [inline packages/connect-web/src/impl/core-in-iframe.ts]: lets replace console.log with  this._log.debug/info

**Review by mroz22 — APPROVED** [review]
good, just some left behind console.logs

**Comment by Hannsek** [comment]
Why auto is not set as a default?

**Comment by martykan** [comment]
To keep the default behavior the same as before, so we don't accidentally break anything. 
The goal is to make it default, but after testing it as opt-in

**Comment by Hannsek** [comment]
I'd push for making it default asap, as I doubt that anyone will change it to auto manually

**Comment by mroz22** [comment]
lets start experimenting with it for the next release
